hazelcast/src/main/java/com/hazelcast/datalink/ReferenceCounter.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.datalink;
+
+import com.hazelcast.core.HazelcastException;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class ReferenceCounter {
+
+    private final AtomicInteger referenceCount = new AtomicInteger(1);
+    private final Callable<Void> destroy;
+
+    public ReferenceCounter(Callable<Void> destroy) {
+        this.destroy = destroy;
+    }
+
+    public ReferenceCounter retain() {
+        int oldCount = referenceCount.getAndIncrement();
+
+        if (oldCount <= 0) {
+            referenceCount.getAndDecrement();
+
+            throw new IllegalStateException("Resurrected a dead object");
+        }
+
+        return this;
+    }
+
+    public boolean release() {
+        long newCount = referenceCount.decrementAndGet();
```

Comment:
You can end up with a negative reference count.

Reply:
We shouldn't - if we do, this class is used incorrectly. I have added a javadoc to make it clearer how this should be used when implementing DataLinks.

Reply:
Sure. But typically you throw the appropriate exception as soon as the problem is detected. If 2 threads at the same time do a release, the problem goes unnoticed.

With the above code, you will not run into this problem.

Reply:
I'm not happy with this solution because it doesn't help out when there is a problem with too frequent releases.